File: T-Bank/Algorithms-and-Data-Structures-2025/6sprint-Graph1/contest/B-Есть_ли_цикл/solution.py
import sys


def dfs(start: int, adj_list: list[list[int]], colors: list[int]) -> bool:
    queue: list[int] = [start]
    while queue:
        vertex: int = queue.pop()
        if not colors[vertex]:
            colors[vertex] = 1
            queue.append(vertex)
            for neighbor in adj_list[vertex]:
                if colors[neighbor] == 1: return True
                elif not colors[neighbor]:
                    queue.append(neighbor)
        else: colors[vertex] = 2
    return False


def solution() -> None:
    """
    Time Complexity O(|V| + |E|)
    Memory Complexity O(|V| + |E|)
    """
    n, m = map(int, sys.stdin.readline().rstrip().split())
    adj_list: list[list[int]] = [[] for _ in range(n + 1)]
    for _ in range(m):
        u, v = map(int, sys.stdin.readline().rstrip().split())
        adj_list[u].append(v)
    colors: list[int] = [0] * (n + 1)
    has_cycle: bool = False
    for vertex in range(1, n + 1):
        if not colors[vertex]:
            has_cycle = has_cycle or dfs(vertex, adj_list, colors)
    sys.stdout.write('{}\n'.format(int(has_cycle)))


if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/6sprint-Graph1/contest/B-Есть_ли_цикл/solution.cpp
#include <iostream>
#include <vector>


bool dfs(int& start, std::vector<std::vector<int>>& adjList, std::vector<int>& colors) {
    std::vector<int> queue = {start};
    while (!queue.empty()) {
        int vertex = queue.back();
        queue.pop_back();
        if (!colors[vertex]) {
            colors[vertex] = 1;
            queue.push_back(vertex);
            for (int neighbor : adjList[vertex]) {
                if (colors[neighbor] == 1) return true;
                if (!colors[neighbor]) queue.push_back(neighbor);
            }
        } else colors[vertex] = 2;
    }
    return false;
}


void solution() {
    /*
    Time Complexity O(|V| + |E|)
    Memory Complexity O(|V| + |E|)
    */
    int n, m;
    std::scanf("%d %d", &n, &m);
    std::vector<std::vector<int>> adjList (n + 1, std::vector<int>());
    for (int i = 0; i < m; ++i) {
        int u, v;
        std::scanf("%d %d", &u, &v);
        adjList[u].push_back(v);
    }
    std::vector<int> colors (n + 1, 0);
    bool hasCycle = false;
    for (int vertex = 1; vertex <= n; ++vertex) {
        if (!colors[vertex]) {
            hasCycle = hasCycle || dfs(vertex, adjList, colors);
        }
    }
    std::printf("%d\n", (hasCycle? 1 : 0));
}


int main() {
    solution();
}